[Dependency Cache] Dependency Cache on CI per Project [without GRADLE_RO_DEP_CACHE]

ParaskP7 opened the pull request:
Project Thread: paaHJt-7CE-p2
Related: #134 + [buildkite-ci#570](https://github.com/Automattic/buildkite-ci/pull/570)
(Softly) Depends On:
- [WCAndroid#13288](https://github.com/woocommerce/woocommerce-android/pull/13288)
- [FluxC#3125](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/3125)
- [ADC#39](https://github.com/Automattic/android-dependency-catalog/pull/39)

---

### Description

This change:

1. Adds a project prefix (`BUILDKITE_PIPELINE_SLUG`) to the existing gradle dependency cache key to enable dependency cache per project. (3b051ba9e5b84e8e2ecfc9cb55047631ef596d83)
2. Enhances [save_cache](https://github.com/Automattic/a8c-ci-toolkit-buildkite-plugin/blob/86bc60f6cf0bb53a7616296dadcfd43aa72c53e9/bin/save_cache) and [restore_cache](https://github.com/Automattic/a8c-ci-toolkit-buildkite-plugin/blob/86bc60f6cf0bb53a7616296dadcfd43aa72c53e9/bin/restore_cache) logging by:
    1. Echoing how long it takes to save/restore the cache. (f777f0207fb0f2f4432a94012ad60fd20a7bedee)
    2. Echoing the (un)compressed cache size during save/restore. (938009b30e9f7afb6230d1650f21489f1cb93976 + fb579e83d220d7d8e2062550963741a00c5fea9d)
3. Replaces the read-only [GRADLE_RO_DEP_CACHE](https://docs.gradle.org/current/userguide/dependency_caching.html#sec:shared-readonly-cache) dependency cache mechanism (using the `$GRADLE_HOME/read-only-dependency-cache` directory), with the more basic [GRADLE_DEP_CACHE](https://docs.gradle.org/current/userguide/dependency_caching.html#sec:cache-copy) dependency cache mechanism (using `$GRADLE_HOME/dependency-cache` directory), for save/restore. (61b1d9d39a46b585ac5cadf67d9d7223fb878ff0)
4. Adds a version suffix (`V2`) to the existing gradle dependency cache key to differentiate between the two, the previous `GRADLE_RO_DEP_CACHE` and the new `GRADLE_DEP_CACHE` dependency cache mechanisms. This was purely done for testing purposes, but I think it might be wise for us to keep this versioning system in place, just in case we update the mechanism again in the future and want to effectively test, then utilize that new change. PS: I am very open to feedback on that. (94d9f7a3829cd21d9f31b5ee962178c26606c71e)

> [!NOTE]
> Enhancing the [save_cache](https://github.com/Automattic/a8c-ci-toolkit-buildkite-plugin/blob/86bc60f6cf0bb53a7616296dadcfd43aa72c53e9/bin/save_cache) and [restore_cache](https://github.com/Automattic/a8c-ci-toolkit-buildkite-plugin/blob/86bc60f6cf0bb53a7616296dadcfd43aa72c53e9/bin/restore_cache) scripts (2.) will positively affect all clients of these scripts, mainly improve things by logging some more information. But, do think if there is going to be a case when these changes might negatively effect some or all those other clients (`I couldn't).

---

### Testing Instructions

This `GRADLE_DEP_CACHE` version of `a8c-ci-toolkit` was thoroughly tested via the below test-only PRs, via the help of the `android-staging` agents (see [buildkite-ci#570](https://github.com/Automattic/buildkite-ci/pull/570)):

- [WCAndroid](https://github.com/woocommerce/woocommerce-android/pull/13170)
- [JP/WPAndroid](https://github.com/wordpress-mobile/WordPress-Android/pull/21550)

To help you assessing that this change is safe use to use and will not cause any trouble for any project, try and follow the below:

- [WCAndroid](https://github.com/woocommerce/woocommerce-android)
  - Latest [save related](https://buildkite.com/automattic/woocommerce-android/builds/26484) build:
    - Prior to all the tasks starting, verify that there [NO](https://buildkite.com/automattic/woocommerce-android/builds/26484#0194503a-4127-47b0-afc3-647aa787a6da/181-182) restore cache triggered (no `local pre-command hook` has run).
    - When all the dependency download tasks finish, [save cache](https://buildkite.com/automattic/woocommerce-android/builds/26484#0194503a-4127-47b0-afc3-647aa787a6da/451-452) get triggered:
      - Using the correct [project prefix and version suffix](https://buildkite.com/automattic/woocommerce-android/builds/26484#0194503a-4127-47b0-afc3-647aa787a6da/451-456).
      - Logging both, [the time it took for the cache to be saved](https://buildkite.com/automattic/woocommerce-android/builds/26484#0194503a-4127-47b0-afc3-647aa787a6da/451-462) and [the size of the saved cache](https://buildkite.com/automattic/woocommerce-android/builds/26484#0194503a-4127-47b0-afc3-647aa787a6da/451-463).
  - Latest [restore related](https://buildkite.com/automattic/woocommerce-android/builds/26485) build:
    - Prior to the `assemble` task starting, [restore cache](https://buildkite.com/automattic/woocommerce-android/builds/26485#0194504e-1c03-41fd-9ed2-a20f6bc350bc/189-190) gets triggered:
      - Using the correct [project prefix and version suffix](https://buildkite.com/automattic/woocommerce-android/builds/26485#0194504e-1c03-41fd-9ed2-a20f6bc350bc/189-193).
      - Logging both, [the time it took for the cache to be restored](https://buildkite.com/automattic/woocommerce-android/builds/26485#0194504e-1c03-41fd-9ed2-a20f6bc350bc/189-198) and [the size of the restored cache](https://buildkite.com/automattic/woocommerce-android/builds/26485#0194504e-1c03-41fd-9ed2-a20f6bc350bc/189-199).
      - Finally, that the restored cache is [placed correctly](https://buildkite.com/automattic/woocommerce-android/builds/26485#0194504e-1c03-41fd-9ed2-a20f6bc350bc/189-201).
    - As the `assemble` tasks starts, verify that there is [NO](https://buildkite.com/automattic/woocommerce-android/builds/26485#0194504e-1c03-41fd-9ed2-a20f6bc350bc/238-248) mention of a `read-only dependency cache` cache.
- [JP/WPAndroid](https://github.com/wordpress-mobile/WordPress-Android)
  - Latest [save related](https://buildkite.com/automattic/wordpress-android/builds/20820) build:
    - Prior to all the tasks starting, verify that there [NO](https://buildkite.com/automattic/wordpress-android/builds/20820#0194503a-caf5-438c-9a8f-f55eaf562078/187-188) restore cache triggered (no `local pre-command hook` has run).
    - When all the dependency download tasks finish, [save cache](https://buildkite.com/automattic/wordpress-android/builds/20820#0194503a-caf5-438c-9a8f-f55eaf562078/528-529) get triggered:
      - Using the correct [project prefix and version suffix](https://buildkite.com/automattic/wordpress-android/builds/20820#0194503a-caf5-438c-9a8f-f55eaf562078/528-533).
      - Logging both, [the time it took for the cache to be saved](https://buildkite.com/automattic/wordpress-android/builds/20820#0194503a-caf5-438c-9a8f-f55eaf562078/528-539) and [the size of the saved cache](https://buildkite.com/automattic/wordpress-android/builds/20820#0194503a-caf5-438c-9a8f-f55eaf562078/528-540).
  - Latest [restore related](https://buildkite.com/automattic/wordpress-android/builds/20821) build:
    - Prior to the `assemble` task starting, [restore cache](https://buildkite.com/automattic/wordpress-android/builds/20821#01945058-1739-4ada-8449-76ed42053121/196-197) gets triggered:
      - Using the correct [project prefix and version suffix](https://buildkite.com/automattic/wordpress-android/builds/20821#01945058-1739-4ada-8449-76ed42053121/196-200).
      - Logging both, [the time it took for the cache to be restored](https://buildkite.com/automattic/wordpress-android/builds/20821#01945058-1739-4ada-8449-76ed42053121/196-205) and [the size of the restored cache](https://buildkite.com/automattic/wordpress-android/builds/20821#01945058-1739-4ada-8449-76ed42053121/196-206).
      - Finally, that the restored cache is [placed correctly](https://buildkite.com/automattic/wordpress-android/builds/20821#01945058-1739-4ada-8449-76ed42053121/196-208).
    - As the `assemble` tasks starts, verify that there is [NO](https://buildkite.com/automattic/wordpress-android/builds/20821#01945058-1739-4ada-8449-76ed42053121/246-262) mention of a `read-only dependency cache` cache.

> [!NOTE]
> The `GRADLE_RO_DEP_CACHE` version of `a8c-ci-toolkit` was also tested via the below test-only PRs:
> - [WCAndroid](https://github.com/woocommerce/woocommerce-android/pull/13122)
> - [JP/WPAndroid](https://github.com/wordpress-mobile/WordPress-Android/pull/21540)

---

### Merge Instructions

- [x] Wait for [buildkite-ci#570](https://github.com/Automattic/buildkite-ci/pull/570) to be reviewed, approved, deployed and merged.
- [x] Wait for [WCAndroid#13288](https://github.com/woocommerce/woocommerce-android/pull/13288) to be reviewed, approved and merged. (`restore cache hook removed`)
- [x] Wait for [FluxC#3125](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/3125) to be reviewed, approved and merged. (`restore cache hook removed`)
- [x] Wait for [ADC#39](https://github.com/Automattic/android-dependency-catalog/pull/39) to be reviewed, approved and merged. (`save cache trigger removed`)
- [x] Remove `do not merge` label.
- [x] Merge this PR.
- [x] Close #98 and #134 test-only PRs.

> [!NOTE]
> These [WCAndroid#13288](https://github.com/woocommerce/woocommerce-android/pull/13288) + [FluxC#3125](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/3125) + [ADC#39](https://github.com/Automattic/android-dependency-catalog/pull/39) PRs are not necessarily needed prior to merging this change, but it is better to clean things up and avoid any confusion with relation to having save/restore working on client/libraries projects, while it is actually not (and shouldn't be as of yet).

---

- [x] I have considered if this change warrants release notes and have added them to the appropriate section in the `CHANGELOG.md` if necessary.

ParaskP7 committed:
Gradle: Measure how long it takes to save/restore the cache

ParaskP7 committed:
Gradle: Measure the (un)compressed cache size during save/restore

ParaskP7 committed:
Gradle: Add project prefix with the standard buildkite pipeline slug

This is done in order to avoid having each client project adding its own
project prefix to distinguish the Gradle dependency cache key.

Test Only PRs:
- WCAndroid: https://github.com/woocommerce/woocommerce-android/
pull/13122
- JP/WPAndroid: https://github.com/wordpress-mobile/WordPress-Android/
pull/21540

For example, for WCAndroid, this would mean that instead of using this
'WOOCOMMERCE' value via this 'GRADLE_DEPENDENCY_CACHE_PROJECT_PREFIX'
environment variable, which ultimately is creating the below
'WOOCOMMERCE_GRADLE_DEPENDENCY_CACHE' key, instead, with this change the
'woocommerce-android_GRADLE_DEPENDENCY_CACHE' key will be
automatically created, taking the 'woocommerce-android' value from the
standard 'BUILDKITE_PIPELINE_SLUG' environment variable.

For more info see: p1734350857095859-slack-C020Q0Z579V

ParaskP7 committed:
Gradle: Add optional cache file to restore cache script

This is done in order for us to be able to output the uncompressed cache
size during cache restoration, just as it is being done during the cache
saving process, for consistency purposes.

But, given the fact that are also other clients of this 'restore_cache'
script, keeping this parameter optional was the preferred approach in
order to not affect, or need to update, any of those other clients of
this specific script.

dangermattic commented:
<!--
  0 Errors
  1 Warning: Please add an entry in the <co...
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Please add an entry in the <code>CHANGELOG.md</code> file to describe the changes made by this PR</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

ParaskP7 committed:
Gradle: Replace gradle ro dependency cache with gradle dependency cache

1. On saving, the script will create a new folder, within the
'GRADLE_HOME' space, named 'dependency-cache', where all the contents
from 'caches/modules-2' will be copied after the build finishes
(minus '*.lock' and 'gc.properties' files, as per the documentation).
2. On restoring, the script will create the 'caches/modules-2' folder,
within the 'GRADLE_HOME' space, and extract the saved contents into it
and before the build starts.
3. On restoring, the script will also check if 'modules-2' directory
exists prior to placing cache. This check is necessary because during
'restore_cache' it might happen that the dependency cache based on this
key is not available for download. Thus, no 'dependency-cache/module-2'
folder will exist to be copied into the 'caches/modules-2' folder, which
will break the script.

ParaskP7 committed:
Gradle: Add v2 suffix on the gradle dependency cache key

This is done in order to differentiate between the 'GRADLE_RO_DEP_CACHE'
and the 'GRADLE_DEP_CACHE' caches, and thus be able to test both caching
strategies in-parallel, without one overwriting the other during save,
or one reusing the other's cache during restore.

Test Only PRs:
- WCAndroid: https://github.com/woocommerce/woocommerce-android/
pull/13170
- JP/WPAndroid: https://github.com/wordpress-mobile/WordPress-Android/
pull/21550

wzieba reviewed (approved):
Looks good to me 🎉

iangmaia commented on bin/restore_cache:
```diff
@@ -1,6 +1,12 @@
 #!/bin/bash -eu
 
 CACHE_KEY=$1
+CACHE_FILE=${2:-} # Optional - defaults to empty if not provided, and the uncompressed cache size not measured (or reported).
```
It's not 100% clear to me what exactly is this file and why would a client want to pass it -- is it just to measure the uncompressed size? Can't we figure this out without requiring a parameter (as we'll uncompress the cache anyway)? 🤔

ParaskP7 replied:
Yea @iangmaia , this is this [$DEP_CACHE_FOLDER_NAME](https://github.com/Automattic/a8c-ci-toolkit-buildkite-plugin/pull/135/files#diff-1848282caeb90d3a6b5fa16fc221cf5b948f4bd8a112fdabc830d23e92297b52R16) folder name, point to this [dependency-cache](https://github.com/Automattic/a8c-ci-toolkit-buildkite-plugin/pull/135/files#diff-1848282caeb90d3a6b5fa16fc221cf5b948f4bd8a112fdabc830d23e92297b52R9) directory.

I couldn't figure out how to make this work so that the uncompressed size is being reported during `restore cache`, just like it is done during `save cache`, but without altering the logic within this [restore_cache](https://github.com/Automattic/a8c-ci-toolkit-buildkite-plugin/blob/94d9f7a3829cd21d9f31b5ee962178c26606c71e/bin/restore_cache) script, me being afraid that me doing so will negatively affect all the other clients of this specific script. Thus, I thought that I better add this optional script argument that will know the directory name of the unzipped file that is downloaded from S3.

Does that make sense? 🤔

PS: I am totally up for doing it better if we can. 🙏

ParaskP7 replied:
Done: 1dda7bd51d20fbb82d5435dc2cbee3b08f5b5153

iangmaia commented on bin/restore_cache:
```diff
@@ -29,11 +36,23 @@ if aws s3api head-object --bucket "$S3_BUCKET_NAME" --key "$CACHE_KEY" > /dev/nu
 	echo "	Decompressing"
 	tar -xf "$CACHE_KEY"
+	if [ -n "$CACHE_FILE" ]; then
+		UNCOMPRESSED_CACHE_SIZE=$(du -sb "$CACHE_FILE" | awk '{print $1}')
```
I've tested this command a bit and `-b` apparently doesn't exist in the macOS' `du` 🤔

ParaskP7 replied:
Oh, interesting, thanks for testing @iangmaia , this comment of yours, plus this [other](https://github.com/Automattic/a8c-ci-toolkit-buildkite-plugin/pull/135#discussion_r1914761899) one, makes me think then, maybe reporting only the compressed cache size is enough (for now). This will then remove a bit of the complexity, right? 😊

iangmaia replied:
Yeah, I mean, we perhaps could spend more time on it to find solutions but I don't see it as such a big deal, so I'd be fine indeed to just remove the uncompressed size report when restoring the cache 🤔

ParaskP7 replied:
Ok, perfect, on it, let's simplify, thanks @iangmaia ! 💯

ParaskP7 replied:
Done: 1dda7bd51d20fbb82d5435dc2cbee3b08f5b5153

ParaskP7 committed:
Gradle: Remove uncompressed cache size measurement

With this change only the compressed cache size is being measured and
echoed simply as 'cache size'. There is no real need to overcomplicate
things by echoing both, the compresses and uncompressed cache size.

iangmaia commented on bin/restore_cache:
```diff
@@ -29,11 +35,17 @@ if aws s3api head-object --bucket "$S3_BUCKET_NAME" --key "$CACHE_KEY" > /dev/nu
 	echo "	Cleaning Up"
 	rm "$CACHE_KEY"
+
+	duration=$SECONDS
```
TIL about `SECONDS` 👍  I always used something like `$(date +%s)`.

iangmaia reviewed (approved):
Looking good 👍
